src/target/riscv/riscv-013.c

```diff
@@ -934,34 +918,39 @@ static int dm013_select_target(struct target *target)
 	return dm013_select_hart(target, info->index);
 }
 
-static int execute_abstract_command(struct target *target, uint32_t command,
-		uint32_t *cmderr)
+static int batch_run_timeout(struct target *target, struct riscv_batch *batch);
+
+static int execute_abstract_command_in_batch(struct target *target,
+		struct riscv_batch *batch, uint32_t command, uint32_t *cmderr)
 {
-	assert(cmderr);
-	*cmderr = CMDERR_NONE;
-	if (debug_level >= LOG_LVL_DEBUG) {
-		switch (get_field(command, DM_COMMAND_CMDTYPE)) {
-			case 0:
-				LOG_DEBUG_REG(target, AC_ACCESS_REGISTER, command);
-				break;
-			default:
-				LOG_TARGET_DEBUG(target, "command=0x%x", command);
-				break;
-		}
-	}
+	riscv_batch_add_dm_write(batch, DM_COMMAND, command, /* read_back */ true,
+			RISCV_DELAY_BASE);
+	riscv_batch_add_dm_read(batch, DM_ABSTRACTCS, RISCV_DELAY_ABSTRACT_COMMAND);
 
-	if (dm_write_exec(target, DM_COMMAND, command, false /* ensure success */) != ERROR_OK)
+	/* Abstract commands are executed while running the batch. */
+	dm013_info_t *dm = get_dm(target);
+	if (!dm)
 		return ERROR_FAIL;
+	dm->abstract_cmd_maybe_busy = true;
 
-	uint32_t abstractcs;
-	int wait_result = wait_for_idle(target, &abstractcs);
-	if (wait_result != ERROR_OK) {
-		/* TODO: can we recover from this? */
-		if (wait_result == ERROR_TIMEOUT_REACHED)
-			LOG_TARGET_DEBUG(target, "command 0x%" PRIx32 " failed (timeout)", command);
-		else
-			LOG_TARGET_DEBUG(target, "command 0x%" PRIx32 " failed (unknown fatal error %d)", command, wait_result);
-		return wait_result;
+	int result = batch_run_timeout(target, batch);
+	if (result != ERROR_OK) {
+		riscv_batch_free(batch);
+		return result;
+	}
+	/* The only read in the batch is of "abstractcs". */
+	uint32_t abstractcs = riscv_batch_get_dmi_read_data(batch, 0);
```

Comment:
This makes strong assumptions about what scans the caller has put into the batch before calling this function.

IMO, this is yet another reason why the split of the responsibility between the functions should be different (please see my other comment).

Reply:
Addressed.